go/tests/api_test.go

```diff
@@ -0,0 +1,233 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"gopkg.in/yaml.v2"
+)
+
+type testFile struct {
+	App   string
+	Tests []test
+}
+
+type test struct {
+	Path string
+	Post any
+	Body any
+}
+
+const hostPort = "http://localhost:3100"
+
+func TestReflectionAPI(t *testing.T) {
+	filenames, err := filepath.Glob(filepath.FromSlash("../../tests/*.yaml"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, fn := range filenames {
+		if filepath.Base(fn) == "pnpm-lock.yaml" {
+			continue
+		}
+		data, err := os.ReadFile(fn)
+		if err != nil {
+			t.Fatal(err)
+		}
+		var file testFile
+		if err := yaml.Unmarshal(data, &file); err != nil {
+			t.Fatal(err)
+		}
+		t.Run(file.App, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			wait, err := startGenkitApp(ctx, file.App)
+			if err != nil {
+				t.Fatal(err)
+			}
+			t.Log("started app")
+			defer func() {
+				cancel()
+				err := wait()
+				t.Logf("wait returned %v", err)
+				// pause("after wait")
+			}()
+			for {
+				time.Sleep(50 * time.Millisecond)
+				t.Log("checking...")
+				if _, err := http.Get(hostPort + "/api/__health"); err == nil {
+					break
+				}
+			}
+			t.Log("app ready")
+			for _, test := range file.Tests {
+				runTest(t, test)
+			}
+		})
+	}
+}
+
+func runTest(t *testing.T, test test) {
+	t.Run(test.Path[1:], func(t *testing.T) {
+		url := hostPort + test.Path
+		var (
+			res *http.Response
+			err error
+		)
+		if test.Post != nil {
+			body, err := json.Marshal(yamlToJSON(test.Post))
+			if err != nil {
+				t.Fatal(err)
+			}
+			res, err = http.Post(url, "application/json", bytes.NewReader(body))
+		} else {
+			res, err = http.Get(url)
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer res.Body.Close()
+		if res.StatusCode != 200 {
+			t.Fatalf("got status %s", res.Status)
+		}
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+		var got any
+		if err := json.Unmarshal(body, &got); err != nil {
+			t.Fatal(err)
+		}
+		want := yamlToJSON(test.Body)
+		msgs := compare(got, want)
+		if len(msgs) > 0 {
+			t.Logf("%s", prettyJSON(got))
+			t.Fatal(strings.Join(msgs, "\n"))
+		}
+	})
+}
+
+func startGenkitApp(ctx context.Context, dir string) (func() error, error) {
+	// If we invoke `go run`, the actual server is a child of the go command, and not
+	// easy to kill. So instead we compile the app and run it directly.
+	tmp := os.TempDir()
+	cmd := exec.Command("go", "build", "-o", tmp, ".")
+	cmd.Dir = dir
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("build: %w", err)
+	}
+
+	cmd = exec.CommandContext(ctx, "./"+dir)
```

Comment:
This looks odd since it looks like you are trying to execute a directory.  This could use a comment.

Reply:
Done.